BuildScript/BuildTasks/MetadataValueDictionary.cs

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+//
+// See the NOTICE file distributed with this work for additional information
+// regarding copyright ownership.  rubicon licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may not use this
+// file except in compliance with the License.  You may obtain a copy of the
+// License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+// License for the specific language governing permissions and limitations
+// under the License.
+//
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Remotion.BuildScript.BuildTasks
+{
+  /// <summary>
+  /// Wraps a <see cref="IDictionary{TKey,TValue}"/> and performs case-insensitive access operations, emulating default msbuild behavior.
+  /// </summary>
+  public class MetadataValueDictionary
+  {
+    private readonly IReadOnlyDictionary<string, string> _rawDictionary;
+
+    public MetadataValueDictionary (IDictionary<string, string> rawDictionary)
+    {
+      _rawDictionary = new Dictionary<string, string> (rawDictionary, StringComparer.OrdinalIgnoreCase);
+    }
+
+    public KeyValuePair<string, string> Single (IEnumerable<string> possibleKeys, Func<Exception> createEmptySequenceException)
+    {
+      var enumeratedPossibleKeys = possibleKeys.ToArray();
+
+      var intersection = _rawDictionary.Keys.Intersect (enumeratedPossibleKeys, StringComparer.OrdinalIgnoreCase).ToArray();
+
+      if (intersection.Length > 1)
+      {
+        var jointPossibleKeys = string.Join (",", enumeratedPossibleKeys);
+        var jointActualKeys = string.Join (",", _rawDictionary.Keys);
+
+        throw new InvalidOperationException ($"Found multiple elements of '[{jointPossibleKeys}]' in '[{jointActualKeys}]'");
+      }
+
+      var key = intersection.SingleOrDefault() ?? throw createEmptySequenceException();
+
+      return new KeyValuePair<string, string> (key, _rawDictionary[key]);
+    }
+
+    public IEnumerable<string> GetKeysForValue (string value)
+    {
+      return _rawDictionary
+          .Where (kvp => string.Equals (kvp.Value, value, StringComparison.OrdinalIgnoreCase))
```

Comment:
If we used (which has other disadvantages, so we don't) Dictionary<K,V> as type of the field, we could use _dictionary.Comparer since that's public on the Dictionary type but not the interface. So, instead we should move the comparer into a backing field and use the throughout this class whereever the comparer is needed explicity. Name should indicate that it's the dictionary's comparer.